techmatters/hrm
# Implement Edit Case Permissions

Primary reviewer @stephenhand 

Notes: 
- This PR is better reviewed commit by commit.
- The `deep-diff` library being used to determine which actions are being performed on a case update assumes that the order is preserved inside arrays. That means, if we eventually try to add a new note, but instead than "pushing" it to the end of the array we drop it in an arbitrary position in between, that will be considered as an edit operation. Clarifying this to make sure we preserve this behavior on the following refactors.

## Description
This PR adds support for "edit case actions". Also done a small refactor around `getActions` helper function in the last two commits. Please let me know if you think those are good or if you prefer me to rollback them.

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1044)

### Verification steps
For this you'll need two counselors (or at least, a valid `token`for two different counselor)
- `npm run start` to start a local server.
- If you don't have any case in your local DB, you need to create one (easier from Flex, after exposing the local server with ngrok and changing `Twilio.Flex.Manager.getInstance().serviceConfiguration.attributes.hrm_base_url`).
- After the case is created, with the creator:
  - Try creating new "items" in the case from the Flex UI (add a note when there are no notes).
  - Try adding more items to an existing property (add a second note).
  - Grab the `token` and try editing some existing stuff (like the last name of a perpetrator). This can be done like
    ```curl -X PUT 'http://localhost:3000/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/cases/<Case ID>' -H "Content-Type: application/json" -H "Authorization: Bearer <the token>" -d '{ "info": { .... the case.info as it is in the DB plus the update  }'```
- Try doing all the sane but with the token of a second counselor. If the configuration in use is `zm` all of them should fail (unauthorized). For the "add" steps you also need to curl (or deactivate the "hide buttons" behavior in the frontend). To do the first you can 
    ```curl -X PUT 'http://localhost:3000/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/cases/<Case ID>' -H "Content-Type: application/json" -H "Authorization: Bearer <the other token>" -d '{ "info": { .... the case.info as it is in the DB plus one more item in one the already existing properties (like a second note)  }'```
